src/erp12/ga_clj/toolbox.cljc

```diff
@@ -25,12 +25,47 @@
   (fn [population]
     (apply min-key by (take size (shuffle population)))))
 
-;; @todo How to handle downsampled lexicase selection? Where do cases come from?
+
+(defn lexicase-selection
+  "Helper for make-lexicase-selection. Takes the population and cases shuffled
+   in a random order. Behaves deterministically to aid in testing."
+  [candidates cases]
+  ;; Stop when cases is empty or candidates has only 1 left.
+  (if (or (empty? cases)
+          (= 1 (count candidates)))
+    (rand-nth candidates)
+    (let [the-case (first cases)
+          best (apply min (map #(nth (:errors %) the-case)
+                               candidates))]
+      (recur (filter #(= best (nth (:errors %) the-case))
+                     candidates)
+             (rest cases)))))
+
+;; @todo Implement epsilon lexicase
+;; Not sure how where I would calculate epsilons just once per generation.
+(defn epsilon-lexicase-selection
+  "Implements semi-dynamic epsilon lexicase, which seems best in trial.
+   semi-dynamic = use local best, but global epsilons that are calculated
+   only once per generation."
+  [candidates cases]
+  :TODO
+  )
+
+;; @todo How to handle downsampled lexicase selection?
+;;                       |--> downsampling I assume would happen wherever individuals
+;;                            are evaluated, not here
+;; @todo Pre-selection filtering of the population? Maybe only if using lexicase-selection?
```

Comment:
I think it is valuable to perform pre-selection on all forms of lexicase because there is nothing stoping the user from enabling epsilon-lexicase on populations with lots of duplicate error vectors.

Reply:
While pre-selection filtering with standard lexicase doesn't change the probabilities of which individuals are selected, it does change the probabilities with some versions of epsilon lexicase, since the candidates pool having duplicates could change the epsilons that get calculated. This is why I excluded epsilon lexicase here.

Though, now that I think about it, I'm pretty sure that the semi-dynamic epsilon lexicase that is recommended might now be changed, since the epsilons are calculated just once per generation. But, I'd want to do some double checking to be sure.

Reply:
Nope, I'm wrong, this should affect all versions of epsilon lexicase. See my post here: https://discourse.pushlanguage.org/t/pre-selection-for-epsilon-lexicase/1466

Because of this, pre-selection can't be applied to epsilon lexicase.

Reply:
Gotcha. Makes sense to me.